keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,925 @@
+
+This command is useful not only for its convenience, but also other visualizations
+may be built upon the raw output if we opt not to support a first-class custom
+template solution in the future.
+
+#### HTML
+
+```shell
+kubectl explain pods --output html
+```
+
+Similarly to [godoc](https://pkg.go.dev), the we suggest to provide a searchable,
+navigable, generated webpage for the kubernetes types of whatever cluster kubectl
+is talking to.
+
+#### Markdown
+
+```shell
+kubectl explain pods --output md
+```
+
+When using the `md` template, a markdown document is printed to stdout, so it
+might be saved and used for a documentation website, for example.
+
+```console
+<TODO>
+```
+
+
+### Risks and Mitigations
+
+#### OpenAPI V3 Not Available
+
+##### Risk
+
+OpenAPI v3 data is not available in the current cluster.
+
+##### Mitigation
+
+If the initial request for OpenAPI V3 fails, the old implementation using
+OpenAPI v2 should be used instead. There is another risk that this fallback process
+takes too long for interactive speed, so timeout should be carefully considered
+to prevent user from waiting too long to see their results. 
+
+
+#### OpenAPI serialization time
+##### Risk
+
+Today there is no interactive-speed way to deserialize protobuf or JSON openapi
+v3 data into the kube-openapi format.
+
+##### Mitigation
+
+There has been recent progress in this area. To unmarshal kube-OpenAPI v3 is now able
+to be done in a performant enough way to do it in the CLI. This KEP's beta release
+should be blocked on the merging of this optimization.
+
+## Design Details
+
+#### Current High-level Approach
+
+TODO: fill with links to referred implementation
+
+1. User types `kubectl explain pods`
+2. kubectl resolves 'pods' to GVR core v1 pods using cluster discovery information
+3. kubectl resolves GVR to its GVK using restmapper
+4. kubectl fetches `/openapi/v2` as protobuf
+5. kubectl parses the protobuf into `gnostic_v2.Document`
+6. kubectl converts `gnostic_v2.Document` into `proto.Models`
+7. kubectl searches the document's `Definitions` for a schema with the
+extension `x-kubernetes-group-version-kind` matching the interested GVK
+8. kubectl renders the definition using its hardcoded printer
+9. If `--recursive` was used, repeat step 8 for the transitive closure of
+  object-typed fields of the top-level object. Concat the results together.
+
+#### Proposed High-level Approach
+
+1. User types `kubectl explain pods`
+2. kubectl resolves 'pods' to GVR core v1 pods using cluster discovery information
+3. kubectl fetches `/openapi/v3/<group>/<version>`
+4. kubectl parses the result as kube-openapi spec3
+5. kubectl locates the return type of the `/apis/<group>/<version>/<resource>` OpenAPI Path
+6. kubectl renders the type using its built-in template for human-readable plaintext
+7. If `--recursive` was used, repeat step 6 for the transitive closure of
+  object-typed fields of the top-level object. Concat the results together.
```

Comment:
Say they are using the default output format with the recursive flag, and we failed to get v3 on the first request for one reason or another. Will we retry on the recursion, or need to track that we shouldn't?

Reply:
The default openapiv2 document a t`/openapi/v2` includes full definitions for every type for all group-versions known to the cluster, so the renderer does not come up for air when rendering. It takes the full spec as input and renders a subset of it. There should be no need to track that the initial call failed.

Similarly, the kubernetes openapiv3 at `/openapi/v3/<group>/<version>`  implementation includes full definitions for every type referenced by the group-version, even from other group-versions.

In general, I think before attempting to render anything, all schemas should be fetched, and if any could not be downloaded, an error should be thrown.